### DISCUSSION TITLE
Purpose of factory methods in reductions for forecasting

### DISCUSSION THREAD

**Discussion by Lovkush-A** [alan-turing-institute/sktime#630] [Q&A] [answered]
What is the purpose of the factory methods at the end of the reduce.py file. E.g `ReducedTimeSeriesRegressionForecaster`?  At least in its current form, they seem redundant. Are there any plans for these in the future?

**Comment by mloning** [comment, marked as answer]
Hi @Lovkush-A, it's mainly for convenience, so that users don't have to remember all the class names. The reduction class names are really wordy too, so the idea was to make it easier with a factory method. 

But I agree, we still need to find a more user-friendly way to create reduction approaches 

Also see the discussion here: https://github.com/sktime/enhancement-proposals/pull/4

**Reply by Lovkush-A** [reply]
I am replying just to make explicit what I think the issues are, for future reference. (As opposed to suggesting you are not aware of the issues)

Using classnames:
DirectRegressionForecaster(...)
RecursiveRegressionForecaster(...)
DirectTimeSeriesRegressionForecaster(...)
RecursiveTimeSeriesRegressionForecaster(...)

Using factory method:
ReducedRegressionForecaster(strategy = 'direct',...)
ReducedRegressionForecaster(strategy = 'recursive',...)
ReducedTimeSeriesRegressionForecaster(strategy = 'direct',...)
ReducedTimeSeriesRegressionForecaster(strategy = 'recursive',...)

'the reduction class names are really wordy too'.  The factory methods' names are (literally!) just as wordy as the class names.

'users dont have to remember all the class names'.  the amount of memorisation involved in factory method is equal to memorisation for class names.

I think we should aim for a single class (which is tunable) like:

```
ReductionForecaster(
    reduced_to = 'time_series' or 'tabular' or ...
    prediction_strategy = 'recursive' or 'direct' or 'multioutput' ...,
```

**Reply by mloning** [reply]
I like the idea of having a single class. A few thoughts: 
* There are multiple reduction approaches though in the time series domain, not all start with forecasting (see figure 1 in our [paper](https://arxiv.org/pdf/1909.07872.pdf)). But we could still have a single reduction class per task (e.g. `ReductionForecaster` as you describe). 
* Reduction approaches may not share hyper-parameters, so exposing them in a single class may be difficult or require users to pass keyword arguments without seeing these arguments in the docstring or method signature. 
* Internally, I think we'd still want to split up implementations of different approaches into multiple classes. The factory method would simply look up and return the appropriate class (e.g. `make_reduction` similar to `make_pipeline` in scikit-learn).

**Reply by fkiraly** [reply]
Yes, I agree with @mloning that having a single class might not be a bad idea, but it is also asking for trouble due to the conditional hyperparameter sets... not sure how one could solve that issue.

My other suggestion would be have a more systematic naming convention, where the start (rather the end) of the class name indicates what the scitype is. This could be combined with shorthands, though it might make the class names less legible.
